alibi_detect/utils/warnings.py

```diff
@@ -0,0 +1,24 @@
+import functools
+import warnings
+
+
+def deprecated_alias(**aliases):
+    def deco(f):
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            rename_kwargs(f.__name__, kwargs, aliases)
+            return f(*args, **kwargs)
+        return wrapper
+    return deco
+
+
+def rename_kwargs(func_name, kwargs, aliases):
+    for alias, new in aliases.items():
+        if alias in kwargs:
+            if new in kwargs:
+                raise TypeError('{} received both {} and {}'.format(
+                    func_name, alias, new))
+            warnings.warn('{} is deprecated; use {}'.format(alias, new),
+                          DeprecationWarning,
+                          3)
```

Comment:
Can you check that the string formatting is performed "early", i.e. there is no earlier output in the stacktrace that contains the non-evaluated string?

Reply:
Couldn't resist testing:
```python
In [1]: def f():
   ...:     raise TypeError('{} received'.format(f.__name__))
   ...: 

In [2]: f()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-c43e34e6d405> in <module>
----> 1 f()

<ipython-input-1-acd52b9b6398> in f()
      1 def f():
----> 2     raise TypeError('{} received'.format(f.__name__))
      3 

TypeError: f received
```

So maybe not the worst output. Also it seems that defining the message first doesn't give much better results:

```python
In [3]: def f():
   ...:     msg = f'{f.__name__} received'
   ...:     raise TypeError(msg)
   ...: 
   ...: 

In [4]: f()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-c43e34e6d405> in <module>
----> 1 f()

<ipython-input-3-ce96ff3445ed> in f()
      1 def f():
      2     msg = f'{f.__name__} received'
----> 3     raise TypeError(msg)
      4 
      5 

TypeError: f received
```

Reply:
Good to know!

Reply:
P.s. I probably should have included an example earlier. This is the result of calling `KSDrift` with `preprocess_x_ref` as a kwarg:

```
<ipython-input-11-12ef60d618f6>:25: DeprecationWarning: preprocess_x_ref is deprecated; use preprocess_at_init
  cd = KSDrift(X_ref, p_val=p_val, preprocess_fn=preprocess_fn, preprocess_x_ref=True)
```